### PR TITLE
GSAPI Integration

### PR DESCRIPTION
**Pull request by msouff**
- brought in and adapted gsapi functions from atcore

### CODE REVIEW

File: tethys_dataset_services/engines/geoserver_engine.py [hunk 1]
```diff
@@ -981,187 +1193,111 @@ def link_sqlalchemy_db_to_geoserver(self, store_id, sqlalchemy_engine, docker=Fa
...
-        new_resource = self.catalog.get_resource(name=table, store=name, workspace=workspace)
-        resource_dict = self._transcribe_geoserver_object(new_resource)
-
-        response_dict = {'success': True,
-                         'result': resource_dict}
-        self._handle_debug(response_dict, debug)
         return response_dict
 
-    def add_table_to_postgis_store(self, store_id, table, debug=False):
+    def create_layer_from_postgis_store(self, store_id, table, debug=False):
```
Comment (swainn) [inline tethys_dataset_services/engines/geoserver_engine.py:1300]: Update example to match new name.

File: tethys_dataset_services/engines/geoserver_engine.py [hunk 2]
```diff
@@ -1238,84 +1374,117 @@ def add_table_to_postgis_store(self, store_id, table, debug=False):
...
-          geometry_srid (string, optional): EPSG spatial reference id of the geometry column. Defaults to 4326.
-          default_style_id (string, optional): Identifier of a style to assign as the default style. Can be a style name or a workspace-name combination (e.g.: "name" or "workspace:name").
-          key_column (string, optional): The name of the key column.
-          parameters (iterable, optional): A list/tuple of tuple-triplets representing parameters in the form (name, default, regex_validation), (e.g.: (('variable', 'pressure', '^[\w]+$'), ('simtime', '0:00:00', '^[\w\:]+$'))
-          debug (bool, optional): Pretty print the response dictionary to the console for debugging. Defaults to False.
-
-        Returns:
-          (dict): Response dictionary
+            layer_id (string): Identifier of the sql view layer to create. Can be a name or a workspace-name combination (e.g.: "name" or "workspace:name").
+            postgis_store_name (string): Identifier of existing postgis store with tables that will be queried by the sql view. (assumption: the datastore belongs to the workspace).
```
Comment (swainn) [inline tethys_dataset_services/engines/geoserver_engine.py:1384]: The workspace should be associated with the store, not the layer.

Change `postgis_store_name` to `store_id` and extract the workspace from that. Update the doc strings too.

Also, change `layer_id` to `layer_name`. Update the doc strings too.

File: tethys_dataset_services/engines/geoserver_engine.py [hunk 3]
```diff
@@ -1238,84 +1374,117 @@ def add_table_to_postgis_store(self, store_id, table, debug=False):
...
+        context = {
+            'workspace': workspace,
+            'feature_name': feature_name,
+            'datastore_name': postgis_store_name,
+            'geoserver_rest_endpoint': self.endpoint,
+            'sql': sql,
+            'geometry_name': geometry_name,
+            'geometry_type': geometry_type,
+            'srid': srid,
+            'parameters': parameters or [],
```
Comment (swainn) [inline tethys_dataset_services/engines/geoserver_engine.py:1411]: `default_style` and `other_styles`?

File: tethys_dataset_services/engines/geoserver_engine.py [hunk 4]
```diff
@@ -1238,84 +1374,117 @@ def add_table_to_postgis_store(self, store_id, table, debug=False):
...
+        with open(gwc_layer_path, 'r') as gwc_layer_file:
+            text = gwc_layer_file.read()
+            template = Template(text)
+            xml = template.render(context)
 
-        # Wrap Up
-        r_feature_layer = self.catalog.get_layer(feature_type_name)
+        retries_remaining = 300
+        while retries_remaining > 0:
+            response = requests.post(
```
Comment (swainn) [inline tethys_dataset_services/engines/geoserver_engine.py:1468]: GSAPI version uses `put`. Why change to `post`?

File: tethys_dataset_services/engines/geoserver_engine.py [hunk 5]
```diff
@@ -2270,40 +2489,106 @@ def delete_store(self, store_id, purge=False, recurse=False, debug=False):
...
+        Delete the specified coverage store.
 
         Args:
-          style_id (string): Identifier of the style to delete.
-          purge (bool, optional): Purge if True.
-          recurse (bool, optional): Delete recursively if True.
-          debug (bool, optional): Pretty print the response dictionary to the console for debugging. Defaults to False.
+            store_id (string): Identifier for the store to be deleted. Can be a store name or a workspace name combination (e.g.: "name" or "workspace:name"). Note that the workspace must be an existing workspace. If no workspace is given, the default workspace will be assigned.
+            workspace (str): the name of the workspace to add the style to
+            name(str):name of the coverage
```
Comment (swainn) [inline tethys_dataset_services/engines/geoserver_engine.py:2499]: Remove these doc strings. Replaced by store_id (typo in the workspace doc string btw)

File: tethys_dataset_services/engines/geoserver_engine.py [hunk 6]
```diff
@@ -2327,3 +2612,214 @@ def validate(self):
...
+            msg = 'Unable to query tile cache status for layer {}:{}. {}:{}'.format(
+                workspace, name, response.status_code, response.text
+            )
+            exception = requests.RequestException(msg, response=response)
+            raise exception
+
+        response_dict = {'success': True, 'result': None}
+        return response_dict 
+
+    def query_tile_cache_tasks(self, workspace, name):
```
Comment (swainn) [inline tethys_dataset_services/engines/geoserver_engine.py:2741]: `workspace, name` -> `layer_id`

File: tethys_dataset_services/engines/geoserver_engine.py [hunk 7]
```diff
@@ -2327,3 +2612,214 @@ def validate(self):
...
+                    status_list.append(dict(temp_dict))
+            return status_list
+        else:
+            msg = 'Unable to terminate tile cache tasks for layer {}:{}. {}:{}'.format(
+                workspace, name, response.status_code, response.text
+            )
+            exception = requests.RequestException(msg, response=response)
+            raise exception
+
+    def enable_time_dimension(self, workspace, coverage_name):
```
Comment (swainn) [inline tethys_dataset_services/engines/geoserver_engine.py:2785]: `workspace, coverage_name` -> `coverage_id`

File: tethys_dataset_services/engines/geoserver_engine.py [hunk 8]
```diff
@@ -2327,3 +2612,214 @@ def validate(self):
         if 'Geoserver Configuration API' not in r.text:
             raise AssertionError('The URL "{0}" is not a valid GeoServer spatial dataset service '
                                  'endpoint.'.format(self.endpoint))
+
+    def modify_tile_cache(self, workspace, name, operation, zoom_start=10, zoom_end=15, grid_set_id=900913,
```
Comment (swainn) [inline tethys_dataset_services/engines/geoserver_engine.py:2616]: `workspace, name` -> `layer_id`

**Review by swainn — CHANGES_REQUESTED** [review]
See comments

File: tethys_dataset_services/engines/geoserver_engine.py (hunk 4 above)
Comment (msouff) [reply tethys_dataset_services/engines/geoserver_engine.py:1489]: This was a bug that Tran had shown me a while ago. The put call returned a warning from geoserver literally saying "use post".

**Review by msouff — COMMENTED** [review]
I believe this is ready.

**Review by swainn — APPROVED** [review]
Looks good.

**Pull request merged**